### PR TITLE
Kilkari CreateSubscription and DeactivateSubscription

### LINKED ISSUE #10: api/kilkari - Create Subscription Request

### LINKED ISSUE #11: api/kilkari - Deactivate Subscription Request

### LINKED ISSUE #32: Kilkari SubscriptionId should be a UUID
According to the API spec, the Kilkari SubscriptionId should be a UUID. It's a String with no validation for now.

### PR DESCRIPTION
**Pull request by llavoie**
Fixes #10, fixes #11, fixes #32

### CODE REVIEW

File: api/src/main/java/org/motechproject/nms/api/web/KilkariController.java [hunk 1]
```diff
@@ -101,4 +104,73 @@ public void saveInboxCallDetails(@RequestBody InboxCallDetailsRequest request) {
...
 
+    /**
+     * 4.2.3
+     * Create Subscription
+     *
+     * IVR shall invoke this API to create a new Kilkari subscription
+     *
+     */
+    @RequestMapping(value = "/subscription",
+            method = RequestMethod.POST,
```
Comment (larubbio) [inline api/src/main/java/org/motechproject/nms/api/web/KilkariController.java]: Should the url be /kilkari/subscription?

File: api/src/main/java/org/motechproject/nms/api/web/KilkariController.java [hunk 2]
```diff
@@ -101,4 +104,73 @@ public void saveInboxCallDetails(@RequestBody InboxCallDetailsRequest request) {
...
+            headers = { "Content-type=application/json" })
+    @ResponseStatus(HttpStatus.OK)
+    public void createSubscription(@RequestBody SubscriptionRequest subscriptionRequest)
+        throws NotFoundException {
```
Comment (larubbio) [inline api/src/main/java/org/motechproject/nms/api/web/KilkariController.java]: Remember Frank said the throws can be removed if NotFoundException extends IllegalStateException

File: api/src/main/java/org/motechproject/nms/api/web/KilkariController.java (hunk 1 above)
Comment (llavoie) [inline api/src/main/java/org/motechproject/nms/api/web/KilkariController.java]: It is. The kilkari part is mapped at the class level.

-----Original Message-----

File: api/src/main/java/org/motechproject/nms/api/web/KilkariController.java [hunk 3]
```diff
@@ -101,4 +104,73 @@ public void saveInboxCallDetails(@RequestBody InboxCallDetailsRequest request) {
...
+    }
+
+    /**
+     * 4.2.4
+     * Deactivate Subscription
+     *
+     * IVR shall invoke this API to deactivate an existing Kilkari subscription
+     *
+     */
+    @RequestMapping(value = "/subscription",
```
Comment (larubbio) [inline api/src/main/java/org/motechproject/nms/api/web/KilkariController.java]: The url is /kilkari/subscription

File: api/src/main/java/org/motechproject/nms/api/web/KilkariController.java (hunk 1 above)
Comment (larubbio) [inline api/src/main/java/org/motechproject/nms/api/web/KilkariController.java]: ok, ignore the next comment that should be coming through the intertubes
right now then.

File: api/src/main/java/org/motechproject/nms/api/web/KilkariController.java (hunk 2 above)
Comment (llavoie) [inline api/src/main/java/org/motechproject/nms/api/web/KilkariController.java]: He's already fixed that in his branch which he hasn't pushed yet, so I figured no need to also do it here.

-----Original Message-----

File: api/src/main/java/org/motechproject/nms/api/web/KilkariController.java [hunk 4]
```diff
@@ -101,4 +104,73 @@ public void saveInboxCallDetails(@RequestBody InboxCallDetailsRequest request) {
...
+            method = RequestMethod.DELETE,
```
Comment (larubbio) [inline api/src/main/java/org/motechproject/nms/api/web/KilkariController.java]: Should you really have a delete method with a post body? :-1: :)

Comment (llavoie) [inline api/src/main/java/org/motechproject/nms/api/web/KilkariController.java]: Following the spec :)

-----Original Message-----

File: api/src/main/java/org/motechproject/nms/api/web/KilkariController.java [hunk 5]
```diff
@@ -101,4 +104,73 @@ public void saveInboxCallDetails(@RequestBody InboxCallDetailsRequest request) {
...
+
+        if (subscription == null) {
+            throw new NotFoundException(String.format(NOT_FOUND, "subscriptionId"));
+        }
+
+        kilkariService.deactivateSubscription(subscription);
+    }
+
+    private boolean validateSubscriptionPack(String name) {
+        return kilkariService.getSubscriptionPack(name) != null;
```
Comment (larubbio) [inline api/src/main/java/org/motechproject/nms/api/web/KilkariController.java]: File this under premature optimization but if you are loading the record just to validate the name, you can use the _count lookup so less data is sent and the object doesn't get constructed.

Comment (llavoie) [inline api/src/main/java/org/motechproject/nms/api/web/KilkariController.java]: Nice!

-----Original Message-----

**Comment by llavoie** [comment]
@larubbio I made the change to use the count lookup for subscriptionpack. Let me know if you have any other feedback.

**Comment by koshalt** [comment]
Looks good to me. (builds & it passes). Leaving it to Rob to merge.

**Comment by larubbio** [comment]
Nooooooooooo!!!!

https://i.imgur.com/1UElQSP.jpg

**Comment by frankhuster** [comment]
Happy to merge it if you're experiencing merge challenges, @larubbio.

**Comment by koshalt** [comment]
http://www.reactiongifs.com/wp-content/uploads/2013/05/kramer-yeah.gif

**Comment by larubbio** [comment]
I did the merge and am running the tests now.

**Pull request merged**